default/python/AI/ResourcesAI.py

```diff
@@ -148,115 +187,338 @@ def check_avoid_change_due_to_bureaucracy(self, pinfo: PlanetFocusInfo, focus: s
+                # is mostly wasted, so ignore it for now.
+                pop = planet.currentMeterValue(fo.meterType.population)
+                pop_gain = potential_pop_increase - planet.habitableSize
+                if pop > pop_gain:
+                    _print_evaluation("would lose more pop (%.1f) than gain everywhere else (%.1f)." % (pop, pop_gain))
+                    continue
 
-        universe.updateMeterEstimates(unbaked_pids)
-        # Protection focus will give the same off-focus Industry and Research targets as Growth Focus
-        for pid, pinfo, planet in planet_infos:
-            po = pinfo.possible_output
-            po[PROTECTION] = po[GROWTH]
-            for focus in [INDUSTRY, RESEARCH, INFLUENCE, PROTECTION]:
-                debug(
-                    "Resource output of planet %s for focus %s is %f/%f/%f",
-                    pinfo.planet,
-                    focus,
-                    po[focus][0],
-                    po[focus][1],
-                    po[focus][2],
+                # If we have a computronium special here, then research focus will have higher priority.
+                if AIDependencies.COMPUTRONIUM_SPECIAL in planet.specials and RESEARCH in planet.availableFoci:
+                    _print_evaluation("has a usable %s" % AIDependencies.COMPUTRONIUM_SPECIAL)
+                    continue
+
+                _print_evaluation(
+                    "considered (pop %.1f, growth gain %.1f, current focus %s)" % (pop, pop_gain, pinfo.current_focus)
                 )
 
+                # add a bias to discourage switching out growth focus to avoid focus change penalties
+                if pinfo.current_focus == GROWTH:
+                    pop -= 4
+
+                ranked_planets.append((pop, pid, planet))
+
+            if not ranked_planets:
+                debug("  --> No suitable location found.")
+                continue
+
+            # sort possible locations by population in ascending order and set population
+            # bonus at the planet with lowest possible population loss.
+            ranked_planets.sort()
+            for pop, pid, planet in ranked_planets:
+                if self.bake_future_focus(pid, GROWTH):
+                    debug("  --> Using growth focus at %s" % planet)
+                    break
+            else:
+                warning("  --> Failed to set growth focus at all candidate locations.")
+
+    def set_planet_production_and_research_specials(self):
+        """Set production and research specials.
+        Sets production/research specials for known (COMPUTRONIUM, HONEYCOMB and CONC_CAMP)
+        production/research specials.
+        Remove planets from list of candidates using bake_future_focus.
+        """
+        # loss should be >= 0 with 0 meaning research is the best focus anyway
+        for loss, pid in sorted(
+            [(self.planet_info[pid].value_below_best(RESEARCH), pid) for pid in computronium_candidates()]
```

Comment:
I'd move this to a variable.  It will take the same 2 lines but will be a bit more readable.